crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,351 @@
+
+    let udp = UdpPacket::new(ip.payload()).ok_or(io::Error::new(
+        io::ErrorKind::InvalidData,
+        "Invalid PQ keys UDP packet received",
+    ))?;
+
+    let mut data = io::Cursor::new(udp.payload());
+
+    let mut version = [0u8; 4];
+    data.read_exact(&mut version)?;
+    let version = u32::from_le_bytes(version);
+    if version != HANDSHAKE_VERSION {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Server responded with invalid PQ handshake version",
+        )
+        .into());
+    }
+
+    let mut ciphertext_len = [0u8; 4];
+    data.read_exact(&mut ciphertext_len)?;
+    let ciphertext_len = u32::from_le_bytes(ciphertext_len);
+
+    if ciphertext_len != CIPHERTEXT_LEN {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Server responded with invalid PQ handshake ciphertext lenght",
+        )
+        .into());
+    }
+
+    data.read_exact(&mut cipherbuf)?;
+
+    let ct = kyber768::Ciphertext::from_bytes(&cipherbuf)
+        .map_err(|err| format!("Invalid PQ ciphertext received: {err:?}"))?;
+
+    Ok(ct)
+}
+
+fn create_get_packet(
+    wg_public: &x25519_dalek::PublicKey,
+    pq_public: &kyber768::PublicKey,
+    rng: &mut impl rand::Rng,
+) -> Vec<u8> {
+    let mut pkgbuf = Vec::with_capacity(1024 * 4); // 4 KiB
+    pkgbuf.resize(IPV4_HEADER_LEN + UDP_HEADER_LEN, 0);
+
+    push_get_method_udp_payload(&mut pkgbuf, wg_public, pq_public);
+    fill_get_packet_headers(pkgbuf.as_mut_slice(), rng);
+
+    pkgbuf
+}
+
+/// The payload looks as follows:
+///
+/// ---------------------------------
+///  version           , u32le , = 1
+/// ---------------------------------
+///  method            , u32le , = 0
+/// ---------------------------------
+///  WG pubkey len     , u32le , = 32
+/// ---------------------------------
+///  WG pubkey bytes   , [u8]
+/// ---------------------------------
+///  Kyber pubkey len  , u32le, = 1184
+/// ---------------------------------
+///  Kyber pubkey bytes, [u8]
+/// ---------------------------------
+fn push_get_method_udp_payload(
+    pkgbuf: &mut Vec<u8>,
+    wg_public: &x25519_dalek::PublicKey,
+    pq_public: &kyber768::PublicKey,
+) {
+    let method = 0u32; // get
+
+    // UDP packet payload
+    pkgbuf.extend_from_slice(&HANDSHAKE_VERSION.to_le_bytes());
+    pkgbuf.extend_from_slice(&method.to_le_bytes());
+    pkgbuf.extend_from_slice(&(wg_public.as_bytes().len() as u32).to_le_bytes());
+    pkgbuf.extend_from_slice(&wg_public.to_bytes());
+    pkgbuf.extend_from_slice(&(pq_public.as_bytes().len() as u32).to_le_bytes());
+    pkgbuf.extend_from_slice(pq_public.as_bytes());
+}
+
+fn fill_get_packet_headers(pkgbuf: &mut [u8], rng: &mut impl rand::Rng) {
+    let pkg_len = pkgbuf.len();
+
+    #[allow(clippy::expect_used)]
+    #[allow(index_access_check)]
+    let mut udppkg = MutableUdpPacket::new(&mut pkgbuf[IPV4_HEADER_LEN..])
+        .expect("UDP buffer should not be too small");
```

Comment:
Please make this compile-time failure instead of runtime

Reply:
How? https://docs.rs/pnet_packet/latest/pnet_packet/udp/struct.MutableUdpPacket.html does not have API with compile time checks

Reply:
Arrrgh, seems like I was distracted yesterday when looking through the code. I was thinking about the `pkgbuf` access, and thought it may be dangerous, but the `expect()` is called upon UDP packet, not the other places

In any case, this would crash the whole tunnel if we pass invalid/unparsable UDP packet, therefore these `expect()` must be handled in a proper way (e.g. by returning an error or similar).